packages/client/src/utils/SynchronizedGraphQLClient.ts

```diff
@@ -0,0 +1,177 @@
+
+export const createWriteContract = <T extends Contract>(
+    address: EthereumAddress,
+    contractInterface: ContractInterface,
+    signer: Signer,
+    name: string,
+    graphQLClient: SynchronizedGraphQLClient
+): ObservableContract<T> => {
+    const contract = withErrorHandlingAndLogging<T>(
+        new Contract(address, contractInterface, signer),
+        name
+    )
+    contract.eventEmitter.on('onTransactionConfirm', (_methodName: string, _tx: ContractTransaction, receipt: ContractReceipt) => {
+        graphQLClient.updateRequiredBlockNumber(receipt.blockNumber)
+    })
+    return contract
+}
+
+// TODO import this from a library (e.g. streamr-test-utils if that is no longer a test-only dependency)
+const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))
+
+const timeout = async (
+    waitTimeMs: number,
+    errorMessage: string,
+    callback: () => void
+): Promise<void> => {
+    await wait(waitTimeMs)
+    callback()
+    throw new Error(errorMessage)
+}
+
+class IndexingState {
+    private blockNumber = 0
+    private gates: Map<number, Gate> = new Map()
+    private getCurrentBlockNumber: () => Promise<number>
+    private pollTimeout: number
+    private pollRetryInterval: number
+    private debug: Debugger
+
+    constructor(
+        getCurrentBlockNumber: () => Promise<number>,
+        pollTimeout: number,
+        pollRetryInterval: number,
+        debug: Debugger
+    ) {
+        this.getCurrentBlockNumber = getCurrentBlockNumber
+        this.pollTimeout = pollTimeout
+        this.pollRetryInterval = pollRetryInterval
+        this.debug = debug
+    }
+
+    async waitUntiIndexed(blockNumber: number): Promise<void> {
+        this.debug(`wait until indexed ${blockNumber}`)
+        const gate = this.getOrCreateGate(blockNumber)
+        await Promise.race([
+            gate.check(),
+            timeout(
+                this.pollTimeout,
+                `timed out while waiting for The Graph index update for block ${blockNumber}`,
+                () => this.removeGate(blockNumber)
+            )
+        ])
+    }
+
+    private getOrCreateGate(blockNumber: number): Gate {
+        let gate: Gate | undefined = new Gate()
+        if (blockNumber > this.blockNumber) {
+            gate = this.gates.get(blockNumber)
+            if (gate === undefined) {
+                const isPolling = this.gates.size > 0
+                gate = new Gate()
+                gate.close()
+                this.gates.set(blockNumber, gate)
+                if (!isPolling) {
+                    this.startPolling()
+                }
+            }
+        }
+        return gate
+    }
+
+    private removeGate(blockNumber: number): void {
+        this.gates.delete(blockNumber)
+    }
+
+    /* eslint-disable no-constant-condition, no-await-in-loop, padding-line-between-statements */
+    private async startPolling(): Promise<void> {
+        this.debug('start polling')
+        while (true) {
```

Comment:
When we come back from `await` it could be the case that remaining gates were removed while we were `await`ing.
 
```suggestion
        while (this.gates.size === 0) {
```